### DISCUSSION TITLE
How to work with dynamic entities?

### DISCUSSION THREAD

**Discussion by kesavkolla** [fastapi/fastapi#6599] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
NA
```

### Description

My model definition is not static eg: I have a table which stores the entity data and another table which stores additionaly attributes of the entity. There is a table which stores the definition of entity.

Here is the example:

Person entity:
id, firstname, lastname, email, address

Person attributes:
id,name,value

Personentitydef
id,name,type

Person table stores data like

--------------------------------------------
id | firstname | lastname | email | address
--------------------------------------------
1 | user1 | bob | user1.bob@domain.org | {"street1": "street1", "city": "junk", "zipcode": "292933"}
2 | user2 | bob | user2.bob@domain.org | {"street1": "street1", "city": "junk", "zipcode": "292933"}

Personentity stores like
--------------------------------------------
id | name | type
--------------------------------------------
1  | gender | text
2  | dateofbirth | date
3  | nationality | text

Personattributes table

--------------------------------------------
id | name | value
--------------------------------------------
1  | "gender" | "male"
1  | "dateofbirth" | "1981-01-01"
1  | "nationality" | "American"
2  | "gender" | "female"
2  | "nationality" | "American"

From business perspective an entity is all the data combined from person table and person attributes table. Idea is attributes do get changed over time. Customers can define their own attributes.

When we provide API for CRUD the logic is query the entity definition table and decide which fields will go in person attributes table and which fields will go into person table directly. As we have all the required metadata in person entity definition table.

Now I am struggling is how can I define the models in FastAPI? I don't know the definition of the model at the coding time. Is it possible to derive the model definitions during request phase maybe in a middle ware etc...? I can cache the entity definitions and can find  a way to invalidate the cache whenever entity definition changes.

Can someone point me in some direction on how to solve this?

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.68.1

### Python Version

Python 3.9.5

**Comment by klaa97** [comment, marked as answer]
I don't think your use-case is natively supported:
- Route's dependencies are static
- OpenAPI is generated and cached, therefore changes in the entity schema wouldn't be propagated

Supporting it should be doable, by generating your own OpenAPI schema; you can follow:
- [This tutorial](https://fastapi.tiangolo.com/advanced/extending-openapi/) to understand how to extend the OpenAPI schema generated by FastAPI and create your own - remember to not cache it though!
- Check out [this file](https://github.com/tiangolo/fastapi/blob/1b6350ad9e67cd700ecec2d360bd06b946c3231c/fastapi/openapi/utils.py#L115) to see how the schema is currently generated; you probably will have to extend it by adding your custom logic (retrieve the schema from the DB, add it in the body/query)

**Comment by tiangolo** [comment]
Thanks for the help here @klaa97  ! 👏  🙇 

Thanks for reporting back and closing the issue @kesavkolla  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by lyutian** [comment]
I'm also have this kind of problem.

Trying to develop a CMDB(lab device management) application, and want the object attributes can be edit by user.
So, need the model/schema to be dynamic.

I found some implement of other framework (like go-restful)  used MongoDB, so the dynamic attribute can be easily added into DB, Seems like NoSQL is the first requirement for this case.
